### PR TITLE
Eliminated redundant file reads to improve RF output

### PR DESCRIPTION
**Pull request by begleysm**
The current example file is capable of continuous output by looping the data in the input file, however, it re-reads the file from memory every time it loops.  During this "read" period there is a full amplitude spike in the DAC output followed by some deadtime before the read is complete and the system can generate RF output again.

To eliminate this problem I modified the example code to read the entire file at once, save it in a buffer, and then loop across that buffer for continuous RF output.  This eliminates the "hiccup" (full amplitude spike and dead time) caused by repeated, unnecessary, reads of the input data file.

See old.png (below) for a plot show the full amplitude spike and dead time associated with a data file re-read.
old

See new.png (below) for a plot showing that this artifact no longer exists once the re-reads are eliminated.
new

### CODE REVIEW

File: host/examples/tx_samples_from_file.cpp [hunk 1]
```diff
@@ -47,19 +47,27 @@ template<typename samp_type> void send_from_file(
     uhd::tx_metadata_t md;
     md.start_of_burst = false;
     md.end_of_burst = false;
```
Comment (mbr0wn) [inline host/examples/tx_samples_from_file.cpp]: This would now be 'true'

File: host/examples/tx_samples_from_file.cpp [hunk 2]
```diff
@@ -47,19 +47,27 @@ template<typename samp_type> void send_from_file(
...
-    while(not md.end_of_burst and not stop_signal_called){
+    //get size of infile
+    infile.seekg (0, std::ios::end);
+    infile_length = infile.tellg()/sizeof(samp_type);
+    infile.seekg (0, std::ios::beg);
 
-        infile.read((char*)&buff.front(), buff.size()*sizeof(samp_type));
-        size_t num_tx_samps = size_t(infile.gcount()/sizeof(samp_type));
+    //create appropriate sized buffer
+    std::vector<samp_type> buff(infile_length);
```
Comment (mbr0wn) [inline host/examples/tx_samples_from_file.cpp]: This could be really huge now.

**Comment by mbr0wn** [comment]
@begleysm Thanks a lot for this effort. There are a couple of issues with your PR, which is why we can't just merge it. First, this will only work with small files -- if you have Gigabytes worth of samples, it'll blow up your memory usage. Also, see the comments I made in the review.

In the examples, we have the other issue that we want to keep them simple, as references, rather than being full-blown applications. If you'd want to send a big file efficiently, you'd implement separate threads for file I/O and device I/O... and then you've developed a full-blown application.

Still, I appreciate your suggestion. Given that people will always look to our examples for how to do things, we really want to make sure they're both useful as well as short and readable. I will take this internal for discussion, but we won't be immediately merging this solution.

Thanks!

**Comment by begleysm** [comment]
Thanks for the detailed replies @mbr0wn.  I understand the decision not to merge.  

As you appear to have understood, the example, as it stands, really doesn't "work" for the 50MB case that I tested.  Having full amplitude spikes and deadtime between transmissions will be problematic for many applications.  I'm not sure if these issues would arise during every continuation of reads from a large file or only at the end of the file when it was reset to the beginning.

Perhaps I'll address your concerns and try to submit a new, separate, example that is "designed for small files" and/or beef up the changes to this example to enable the multi-threading required to keep the DAC output smooth & continuous.

**Comment by mbr0wn** [comment]
Thanks for your comments. I would encourage you to discuss this on usrp-users, too. One thing that is really important for this specific issue is that the example stays concise and not overladen with features, for sake of readability.

**Pull request closed without merging**